src/base.cpp

```diff
@@ -47,12 +47,13 @@
 
 namespace lbann {
 
-lbann_comm* initialize(int& argc, char**& argv, int seed) {
+lbann_comm_ptr initialize(int& argc, char**& argv, int seed) {
   // Initialize Elemental.
   El::Initialize(argc, argv);
   // Create a new comm object.
   // Initial creation with every process in one model.
-  auto* comm = new lbann_comm(0);
+  auto comm = lbann_comm_ptr{new lbann_comm(0), &lbann::finalize };
```

Comment:
It bothers me that the user calls `initialize` without also calling `finalize`. It seems to me that we're essentially making `lbann_comm` a singleton object. Going further down this path, we would put `initialize` inside `lbann_comm`'s constructor and `finalize` in the destructor. This has it's own weirdness - the user must make sure it's lifespan matches the beginning and end of the program. We may also run into the case where someone might want multiple instances of `lbann_comm`, but this is not too important.

Reply:
I very deliberately did not put the `initialize` in the Ctor nor `finalize` in the dtor. That is *not* what we want to do. This makes `'lbann_comm_ptr` an _ad hoc_ RAII wrapper for MPI since `initialize` is only called in scopes in which the caller is responsible for freeing the MPI resources. The `lbann_comm` that gets returned from `initialize` is really "COMM_WORLD"; when the world blows up, nothing lives on anyway.

The "proper" fix for this is to actually RAII the interface. We should create a class `lbann::session` or something that calls `initialize()` in its ctor (i.e., when it acquires MPI/CUDA/whatever resources) and 'finalize()` in its dtor (i.e., when it returns those resources). This should be completely distinct from the `lbann_comm` object. Indeed, it will be an error to try to create an `lbann_comm` if you're not nested inside a scope in which an `lbann::session()` does not exist.

Reply:
Just to clarify about the "singleton" comment: Note that `lbann_comm_ptr` is *not* wrapping every `lbann_comm` object -- only the one returned from `initialize' is managed this way. This does *not* make the `lbann_comm` object a singleton by any means. Logically, the thing that gets returned from `intialize` is "COMM_WORLD", which is "an object" that must "exist" (quotes cuz handle to opaque blahblahblah C mumbojumbo) for the entire valid MPI session. If this object gets deleted, logically the session has ended. It's now time to `finalize()`. Indeed, you can split that thing up all you want into sub-`lbann_comm` objects and this is semantically valid (and encouraged).